rat/compiler.py

```diff
@@ -1,488 +1,501 @@
-        integer_shifts: List[int] = []
-        for shift_expr in subscript.shifts:
-            match shift_expr:
-                case ast.IntegerConstant():
-                    integer_shifts.append(shift_expr.value)
-                case _:
-                    try:
-                        shift_code_generator = codegen_backends.BaseCodeGenerator(self.variable_table)
-                        shift_code_generator.generate(shift_expr)
-                        folded_shift = int(eval(shift_code_generator.get_expression_string()))
-                        integer_shifts.append(folded_shift)
-                    except Exception as e:
-                        error_msg = f"Failed evaluating shift. Shift amount expressions must be an expression which can be evaluated at compile-time."
-                        raise CompileError(error_msg, shift_expr.range) from e
-
-        # If there is a non-zero shift, set pad needed to True
-        pad_needed = any(shift != 0 for shift in integer_shifts)
-
-        return tuple(integer_shifts), pad_needed
-
-    def build_variable_table(self):
-        """
-        Builds the "symbol table", which holds information for all variables in the model.
-        """
-        self.variable_table = VariableTable(self.data)
-
-        # Add entries to the symbol table for all data/params
-        for top_expr in self.expr_tree_list:
-            for symbol in top_expr:
+                        return
+
+                record = ConstantVariableRecord(node.name, argument_count)
+
+                try:
+                    record.bind(subscript_names, self.data)
+                except KeyError as e:
+                    msg = f"{node.name} is in control flow and must bind to input data but that failed"
+                    raise CompileError(msg, node) from e
+                except Exception as e:
+                    msg = f"Irrecoverable error binding {node.name} in control flow to data"
+                    raise CompileError(msg, node) from e
+            else:
+                record = ConstantVariableRecord(node.name, argument_count)
+                try:
+                    record.bind(subscript_names, self.data)
+                except KeyError:
+                    record = SampledVariableRecord(node.name, argument_count)
+                except Exception as e:
+                    msg = f"Irrecoverable error binding {node.name} to data"
+                    raise CompileError(msg, node) from e
+
+            self.variable_table[node.name] = record
+        else:
+            record = self.variable_table[node.name]
+
+            if isinstance(record, ConstantVariableRecord):
+                if argument_count == 0:
+                    if node.name not in self.primary_node_and_subscript_names:
+                        msg = (
+                            f"{node.name} is a constant variable and must be a primary variable subscript if it appears with no subscripts"
```

Comment:
Oh, so don't actually traverse if statements here?

I think we'd still need the in_control_flow logic but breaking this into two seems fine.

Reply:
We so need to traverse through if statements. But since it goes through the arglist of variables, it flags `in_control_flow` to true even for variables that aren't in predicates of ifelse.